thunderbirdparser/src/org/sleuthkit/autopsy/thunderbirdparser/EmailMessageThreader.java

```diff
@@ -0,0 +1,674 @@
+            }
+
+            // If both containers are dummy/empty append the children of one to the other
+            if (tableContainer.getMessage() == null && rootSetContainer.getMessage() == null) {
+                tableContainer.addChildren(rootSetContainer.getChildren());
+                rootSetContainer.clearChildren();
+                continue;
+            }
+
+            // one container is empty, but the other is not, make the non-empty one be a
+            // child of the empty
+            if ((tableContainer.getMessage() == null && rootSetContainer.getMessage() != null)
+                    || (tableContainer.getMessage() != null && rootSetContainer.getMessage() == null)) {
+
+                if (tableContainer.getMessage() == null) {
+                    tableContainer.addChild(rootSetContainer);
+
+                } else {
+                    rootSetContainer.addChild(tableContainer);
+                    subject_table.remove(rootSubject, tableContainer);
+                    subject_table.put(rootSubject, rootSetContainer);
+
+                    finalSet.add(rootSetContainer);
+                }
+
+                continue;
+            }
+
+            // tableContainer is non-empty and it's message's subject does not begin
+            // with 'RE:' but rootSetContainer's message does begin with 'RE:', then
+            // make rootSetContainer a child of tableContainer
+            if (tableContainer.getMessage() != null
+                    && !tableContainer.isReplySubject()
+                    && rootSetContainer.isReplySubject()) {
+                tableContainer.addChild(rootSetContainer);
+                continue;
+            }
+
+            // If table container is non-empy, and table container's subject does 
+            // begin with 'RE:', but rootSetContainer does not start with 'RE:'
+            // make tableContainer a child of rootSetContainer
+            if (tableContainer.getMessage() != null
+                    && tableContainer.isReplySubject()
+                    && !rootSetContainer.isReplySubject()) {
+                rootSetContainer.addChild(tableContainer);
+                subject_table.put(rootSubject, rootSetContainer);
+                finalSet.add(rootSetContainer);
+                continue;
+            }
+
+            // rootSetContainer and tableContainer either both have 'RE' or
+            // don't.  Create a new dummy container with both containers as
+            // children.
+            Container newParent = new Container();
+            newParent.addChild(tableContainer);
+            newParent.addChild(rootSetContainer);
+            subject_table.remove(rootSubject, tableContainer);
+            subject_table.put(rootSubject, newParent);
+
+            finalSet.add(newParent);
+        }
+        return finalSet;
+    }
+
+    /**
+     * Creates a Hashtable of Container and subjects. There will be one
+     * Container subject pair for each unique subject.
```

Comment:
We should document here what happens when there are groups with the same subject.  Which one get saved?

Reply:
At this point in the code it doesn't matter which one gets saved. But it would be the first one we find.  This is just a list of unique subjects.

Reply:
I put a big comment before the add.